pkg/streamingpromql/engine_test.go

```diff
@@ -1682,3 +1702,133 @@ func TestAnnotations(t *testing.T) {
+	// These will be used for binary operations, so we can add up to 3 series.
+	labelCombinations := combinations(labels, 2)
+	labelCombinations = append(labelCombinations, combinations(labels, 3)...)
+
+	expressions := []string{}
+
+	// Binary operations
+	for _, labels := range labelCombinations {
+		if len(labels) >= 2 {
+			for _, op := range []string{"+", "-", "*", "/"} {
+				binaryExpr := fmt.Sprintf(`series{label="%s"}`, labels[0])
+				for _, label := range labels[1:] {
+					binaryExpr += fmt.Sprintf(` %s series{label="%s"}`, op, label)
+				}
+				expressions = append(expressions, binaryExpr)
+			}
+		}
+	}
+
+	// For aggregations, also add combinations of 4 labels. (e.g., "a,b,c,d", "c,d,e,f" etc)
+	labelCombinations = append(labelCombinations, combinations(labels, 4)...)
+
+	for _, labels := range labelCombinations {
+		labelRegex := strings.Join(labels, "|")
+		// Aggregations
+		for _, aggFunc := range []string{"sum", "avg", "min", "max"} {
+			expressions = append(expressions, fmt.Sprintf(`%s(series{label=~"(%s)"})`, aggFunc, labelRegex))
+			expressions = append(expressions, fmt.Sprintf(`%s by (group) (series{label=~"(%s)"})`, aggFunc, labelRegex))
+			expressions = append(expressions, fmt.Sprintf(`%s without (group) (series{label=~"(%s)"})`, aggFunc, labelRegex))
+		}
+		expressions = append(expressions, fmt.Sprintf(`rate(series{label=~"(%s)"}[45s])`, labelRegex))
+		expressions = append(expressions, fmt.Sprintf(`avg(rate(series{label=~"(%s)"}[2m15s]))`, labelRegex))
+	}
+
+	timeRanges := []struct {
+		loadStep int
+		interval time.Duration
+	}{
+		{loadStep: 1, interval: 1 * time.Minute},
+		{loadStep: 1, interval: 6 * time.Minute},
+		{loadStep: 1, interval: 5 * time.Minute},
+		{loadStep: 6, interval: 6 * time.Minute},
+		{loadStep: 6, interval: 5 * time.Minute},
+	}
+
+	for _, tr := range timeRanges {
+		start := timestamp.Time(0)
+		end := start.Add(time.Duration(pointsPerSeries*tr.loadStep) * time.Minute) // Deliberately queries 1 step past the final loaded point
+
+		storage := promqltest.LoadedStorage(t, fmt.Sprintf("load %dm", tr.loadStep)+samples)
+		t.Cleanup(func() { require.NoError(t, storage.Close()) })
+
+		for _, expr := range expressions {
+			testName := fmt.Sprintf("Expr: %s, Start: %d, End: %d, Interval: %s", expr, start.Unix(), end.Unix(), tr.interval)
+			t.Run(testName, func(t *testing.T) {
+				q, err := prometheusEngine.NewRangeQuery(context.Background(), storage, nil, expr, start, end, tr.interval)
+				require.NoError(t, err)
+				defer q.Close()
+				expectedResults := q.Exec(context.Background())
+
+				q, err = mimirEngine.NewRangeQuery(context.Background(), storage, nil, expr, start, end, tr.interval)
+				require.NoError(t, err)
+				defer q.Close()
+				mimirResults := q.Exec(context.Background())
+
+				// We currently omit checking the annotations due to a difference between the engines.
+				// This can be re-enabled once https://github.com/prometheus/prometheus/pull/14910 is vendored.
```

Comment:
As mentioned offline - given https://github.com/prometheus/prometheus/pull/14910 has been merged, could we cherry-pick it into mimir-prometheus and then vendor that into Mimir so we don't need these workarounds without needing to do the pending Prometheus 3.0 merge?

Reply:
Lets deal with this separately as I think it's a bigger and more complicated task and we have a way forward here for now.